src/core/main.lua

```diff
@@ -44,10 +45,30 @@ function main ()
       print("unsupported program: "..program:gsub("_", "-"))
       usage(1)
    else
-      require(modulename(program)).run(args)
+      -- Fork into worker process and supervisor
+      local worker_pid = S.fork()
+      if worker_pid == 0 then
+         -- Worker
+         S.prctl("set_pdeathsig", "hup")
+         require(modulename(program)).run(args)
+      else
+         -- Supervisor
+         local exit_signals = "hup, int, quit, term, chld"
+         local signalfd = S.signalfd(exit_signals)
+         S.sigprocmask("block", exit_signals)
+         local signals, err = S.util.signalfd_read(signalfd)
+         assert(signals, tostring(err))
+         if not signals[1].chld then S.kill(worker_pid, "kill") end
+         shutdown(S.getpid())
```

Comment:
It is a mystery to me why this works. E.g. I had assumed I needed to pass `worker_pid` to `shutdown` since the worker goes on to create the `/var/run` directory based on `S.getpid()`, but apparently when the worker calls `S.getpid()` it returns the parent pid? _confused_